compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -14117,6 +14117,168 @@ bool isSettingOp(uint8_t byteValue, TR::InstOpCode::Mnemonic SI_opcode)
    }
 
 
+
+TR::Register *
+OMR::Z::TreeEvaluator::bitpermuteEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+
+   TR_ASSERT(node->getNumChildren() == 3, "Wrong number of children in bitpermuteEvaluator");
+   TR::Node *value = node->getChild(0);
+   TR::Node *addr = node->getChild(1);
+   TR::Node *length = node->getChild(2);
+
+   bool nodeIs64Bit = node->getSize() == 8;
+   auto valueReg = cg->evaluate(value); // this will either be a constant or a load
+   auto addrReg = cg->evaluate(addr);   // this will either be a constant or a load
+
+   TR::Register *tmpReg = cg->allocateRegister(TR_GPR);
+
+   //Zero result reg
+   TR::Register *resultReg = cg->allocateRegister(TR_GPR);
+   generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, resultReg, resultReg);
+
+   if (length->getOpCode().isLoadConst())
+      {
+      // Manage the constant length case
+      uintptrj_t arrayLen = length->getIntegerNodeValue<uintptrj_t>();
+      for (uintptrj_t x = 0; x < arrayLen; ++x)
+         {
+         // zero tmpReg
+         generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, tmpReg, tmpReg);
+
+         //load the shift amount into tmpReg
+         //TODO: Check to see if we need sourceMR->stopUsingMemRefRegister(cg)
+         TR::MemoryReference *sourceMR = generateS390MemoryReference(addrReg, x, cg, NULL);
+         generateRXYInstruction(cg, TR::InstOpCode::LB, node, tmpReg, sourceMR);
+
+         if (!(node->getDataType() == TR::Int32 || node->getDataType() == TR::Int16 || node->getDataType() == TR::Int8 || node->getDataType() == TR::Int64))
+            {
+            TR_ASSERT(0, " Unsupported DataType for bitPermute op");
+            }
+         if (node->getDataType() != TR::Int64)
+            {
+            // Java Spec defines the shift amount for integer shifts to be masked with
+            // 0x1f (0-31). For long shifts, the shift instruction will only consider
+            // the first 6 bits of the operand as the shift amount. So no mask is required in that case
+            generateRIInstruction(cg, TR::InstOpCode::NILL, node, tmpReg, 0x1f);   
+            //TODO: There should be two of these
+            }
+         //create memory reference using tmpReg (shift amount), then shift valueReg by by shift amount
```

Comment:
> by by

Please remove the duplication and insert space after `//` and capitalize the first word.

Reply:
Done